compiler/src/iree/compiler/Dialect/Stream/IR/StreamOpFolders.cpp

```diff
@@ -2331,6 +2333,518 @@ void AsyncConcurrentOp::getCanonicalizationPatterns(RewritePatternSet &results,
+// Only succeeds if single-use chain: producer -> await -> consumer.
+// This allows folding patterns to work with legal IR that properly awaits
+// timeline resources before consuming them.
+static Value tryLookThroughAwait(Value operand) {
+  auto awaitOp = operand.getDefiningOp<IREE::Stream::TimepointAwaitOp>();
+  if (!awaitOp)
+    return operand;
+  if (!operand.hasOneUse())
+    return operand;
+
+  // Find awaited resource (await maps timepoint + resources -> resources).
+  for (auto [resource, result] :
+       llvm::zip_equal(awaitOp.getResourceOperands(), awaitOp.getResults())) {
+    if (result == operand) {
+      // Check resource comes from timeline op with single use.
+      if (resource.hasOneUse() &&
+          isa_and_nonnull<IREE::Stream::TimelineOpInterface>(
+              resource.getDefiningOp())) {
+        return resource;
+      }
+      break;
+    }
+  }
+  return operand;
+}
+
+// For result folding: finds consumer of timeline result through await barrier.
+// Returns {consumer_op, consumed_value} if single-use chain exists.
+static std::pair<Operation *, Value>
+findConsumerThroughAwait(Value timelineResult) {
+  if (!timelineResult.hasOneUse())
+    return {nullptr, nullptr};
+
+  Operation *consumer = *timelineResult.getUsers().begin();
+
+  // If consumer is await, look through it to find actual consumer.
+  if (auto awaitOp = dyn_cast<IREE::Stream::TimepointAwaitOp>(consumer)) {
+    for (auto [resource, result] :
+         llvm::zip_equal(awaitOp.getResourceOperands(), awaitOp.getResults())) {
+      if (resource == timelineResult) {
+        if (!result.hasOneUse())
+          return {nullptr, nullptr};
+        return {*result.getUsers().begin(), result};
+      }
+    }
+    return {nullptr, nullptr};
+  }
+
+  return {consumer, timelineResult};
+}
+
+// Folds resource.subview on the load result into the load operation.
+// This adjusts the parameter offset and size to load only the required portion.
+//
+// Example:
+//  %loaded = stream.async.parameter.load "scope"::"key"[%offset]
+//      : !stream.resource<constant>{%size}
+//  %view = stream.resource.subview %loaded[%view_offset]
+//      : !stream.resource<constant>{%size} ->
+//      !stream.resource<constant>{%view_size}
+// ->
+//  %new_offset = arith.addi %offset, index_cast(%view_offset)
+//  %loaded = stream.async.parameter.load "scope"::"key"[%new_offset]
+//      : !stream.resource<constant>{%view_size}
+struct FoldAsyncParameterLoadResultSubview
+    : public OpRewritePattern<AsyncParameterLoadOp> {
+  using Base::Base;
+  LogicalResult matchAndRewrite(AsyncParameterLoadOp op,
+                                PatternRewriter &rewriter) const override {
+    // Find consumer through optional await barrier (single-use chain).
+    auto [consumer, consumedValue] = findConsumerThroughAwait(op.getResult());
+    auto subviewOp =
+        dyn_cast_or_null<IREE::Stream::ResourceSubviewOp>(consumer);
```

Comment:
style nit: prefer `dyn_cast_if_present`. I made a cleanup change recently, fyi: https://github.com/iree-org/iree/commit/33db935a33ddab6a8358418c0ae6dc263e56b0bb


```suggestion
    auto subviewOp =
        dyn_cast_if_present <IREE::Stream::ResourceSubviewOp>(consumer);
```